test/src/unit-filter.cc

```diff
@@ -0,0 +1,321 @@
+/**
+ * @file unit-filter-buffer.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2017-2021 TileDB, Inc.
+ * @copyright Copyright (c) 2016 MIT and Intel Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Tests the `FilterBuffer` class.
+ */
+
+#include "test/src/helpers.h"
+#include "tiledb/common/status.h"
+#include "tiledb/sm/array_schema/array_schema.h"
+#include "tiledb/sm/array_schema/dimension.h"
+#include "tiledb/sm/array_schema/domain.h"
+#include "tiledb/sm/buffer/buffer.h"
+#include "tiledb/sm/config/config.h"
+#include "tiledb/sm/crypto/encryption_key.h"
+#include "tiledb/sm/enums/compressor.h"
+#include "tiledb/sm/enums/datatype.h"
+#include "tiledb/sm/enums/encryption_type.h"
+#include "tiledb/sm/enums/filter_option.h"
+#include "tiledb/sm/enums/filter_type.h"
+#include "tiledb/sm/filter/bit_width_reduction_filter.h"
+#include "tiledb/sm/filter/bitshuffle_filter.h"
+#include "tiledb/sm/filter/byteshuffle_filter.h"
+#include "tiledb/sm/filter/checksum_md5_filter.h"
+#include "tiledb/sm/filter/checksum_sha256_filter.h"
+#include "tiledb/sm/filter/compression_filter.h"
+#include "tiledb/sm/filter/encryption_aes256gcm_filter.h"
+#include "tiledb/sm/filter/noop_filter.h"
+#include "tiledb/sm/filter/positive_delta_filter.h"
+
+#include <catch.hpp>
+#include <iostream>
+
+using namespace tiledb::sm;
+
+TEST_CASE(
+    "Filter: Test bit width reduction filter deserialization",
+    "[filter][bit-width-reduction]") {
+  uint32_t max_window_size0 = 1024;
+  BitWidthReductionFilter filter0(max_window_size0);
+  Buffer buffer;
+  filter0.serialize(&buffer);
+
+  ConstBuffer constbuffer(&buffer);
+  auto&& [st_filter, filter1]{Filter::deserialize(&constbuffer)};
+  CHECK(st_filter.ok());
```

Comment:
`REQUIRE` is appropriate here. If deserialize did not succeed, the remainder of the tests are moot.